File: grpc-java-examples/src/main/java/com/example/grpc/methodtypes/streaming/server/ServerStreamingServer.java
package com.example.grpc.methodtypes.streaming.server;

import com.example.grpc.EchoRequest;
import com.example.grpc.EchoResponse;
import com.example.grpc.EchoServiceGrpc;
import com.example.grpc.Servers;
import io.grpc.stub.StreamObserver;

import java.util.logging.Level;
import java.util.logging.Logger;

public class ServerStreamingServer {

    private static final Logger logger = Logger.getLogger(CancellationServerStreamingServer.class.getName());

    public static void main(String[] args) throws Exception {
        Servers.start(new EchoServiceImpl());
    }

    private static class EchoServiceImpl extends EchoServiceGrpc.EchoServiceImplBase {
        @Override
        public void serverStreamingEcho(EchoRequest request, StreamObserver<EchoResponse> responseObserver) {
            var name = request.getMessage();
            logger.log(Level.INFO, "request: {0}", name);

            responseObserver.onNext(EchoResponse.newBuilder().setMessage("hello " + name).build());
            responseObserver.onNext(EchoResponse.newBuilder().setMessage("guten tag " + name).build());
            responseObserver.onNext(EchoResponse.newBuilder().setMessage("bonjour " + name).build());
            responseObserver.onCompleted();
        }
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/methodtypes/streaming/server/ServerStreamingServer.kt
package com.example.grpc.methodtypes.streaming.server

import com.example.grpc.EchoRequest
import com.example.grpc.EchoResponse
import com.example.grpc.EchoServiceGrpcKt
import com.example.grpc.Servers
import com.example.grpc.echoResponse
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.util.logging.Logger

object ServerStreamingServer {
  private val logger = Logger.getLogger(ServerStreamingServer::class.java.name)

  @JvmStatic
  fun main(args: Array<String>) {
    Servers.start(EchoServiceImpl())
  }

  private class EchoServiceImpl : EchoServiceGrpcKt.EchoServiceCoroutineImplBase() {
    override fun serverStreamingEcho(request: EchoRequest): Flow<EchoResponse> {
      val name = request.message
      logger.info("request: $name")
      return flow {
        emit(echoResponse { message = "hello $name" })
        emit(echoResponse { message = "guten tag $name" })
        emit(echoResponse { message = "bonjour $name" })
      }
    }
  }
}